client/session/grpctransport/grpc.go

```diff
@@ -0,0 +1,172 @@
+	matches := getMethodRe().FindAllStringSubmatch(in, -1)
+	if len(matches) != 1 || len(matches[0]) != 3 {
+		return "", "", errors.Errorf("invalid method name %s", in)
+	}
+	return matches[0][1], matches[0][2], nil
+}
+
+func (gt *grpcTransportFactory) NewHandler(ctx context.Context, conn net.Conn) (session.TransportHandler, error) {
+	return newHandler(ctx, conn)
+}
+func (gt *grpcTransportFactory) NewCaller(ctx context.Context, conn net.Conn) (session.TransportCaller, error) {
+	return newCaller(ctx, conn)
+}
+
+type callbackSelector struct {
+	id     string
+	method string
+}
+
+type grpcHandler struct {
+	conn     net.Conn
+	handlers map[callbackSelector]session.HandleFunc
+}
+
+func newHandler(ctx context.Context, conn net.Conn) (*grpcHandler, error) {
+	gh := &grpcHandler{
+		conn:     conn,
+		handlers: make(map[callbackSelector]session.HandleFunc),
+	}
+	return gh, nil
+}
+
+func (gh *grpcHandler) Register(id, method string, f session.HandleFunc) error {
+	// type HandleFunc func(ctx context.Context, opts map[string][]string, stream Stream) error
+	gh.handlers[callbackSelector{id: id, method: method}] = f
+	return nil
+}
+
+func (gh *grpcHandler) Serve(ctx context.Context) error {
+	gs := grpc.NewServer()
+	go func() {
+		<-ctx.Done()
+		gh.conn.Close()
+	}()
+
+	services := make(map[string][]callbackSelector)
+
+	for sel := range gh.handlers {
+		sname := serviceName(sel.id, sel.method)
+		services[sname] = append(services[sname], sel)
+	}
+
+	for name, selectors := range services {
+		var streams []grpc.StreamDesc
+		for _, sel := range selectors {
+			func(sel callbackSelector) {
+				streams = append(streams, grpc.StreamDesc{
+					StreamName: streamName(sel.id, sel.method),
+					Handler: func(srv interface{}, stream grpc.ServerStream) error {
+						logrus.Debugf("handling %v", sel)
+						var s session.Stream
+						ctx := stream.Context()
+						md, _ := metadata.FromContext(ctx)
+						s = stream
+						return gh.handlers[sel](ctx, md, &s)
+					},
+					ServerStreams: true,
+					ClientStreams: true,
+				})
+			}(sel)
+		}
+		gs.RegisterService(&grpc.ServiceDesc{
+			ServiceName: name,
+			HandlerType: (*interface{})(nil),
+			Methods:     []grpc.MethodDesc{},
+			Streams:     streams,
+		}, (*interface{})(nil))
+	}
+	logrus.Debugf("serving")
+	(&http2.Server{}).ServeConn(gh.conn, &http2.ServeConnOpts{Handler: gs})
+	return nil
+}
+
+type grpcCaller struct {
+	cc *grpc.ClientConn
+}
+
+func newCaller(ctx context.Context, conn net.Conn) (*grpcCaller, error) {
+	dialOpt := grpc.WithDialer(func(addr string, d time.Duration) (net.Conn, error) {
+		return conn, nil
+	})
+
+	cc, err := grpc.DialContext(ctx, "", dialOpt, grpc.WithInsecure())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create grpc client")
+	}
+
+	gc := &grpcCaller{
+		cc: cc,
+	}
+
+	return gc, nil
+}
+
+func (gc *grpcCaller) Call(ctx context.Context, id, method string, opt map[string][]string) (session.Stream, error) {
```

Comment:
In a user code perspective, the transport API feels a bit too complex anyway. On the exposing side, the handler registration code seems overly complex (having to define a method, that takes a callback, that you must call for each method you want to expose feels so 80's in a world with introspection-enabled languages).
On the consuming side, having to call methods by putting their name in a string, passing arguments as a combination of a map of string slices and object streaming is much lower level that I would expect from a modern transport stack.

Ultimately what I want is a way to expose a simple request/response interface implementation like 
```
type AuthDelegate interface{
 GetBasicCredentials(registry string)(username, password string, err error)
 GetOAuthAccessToken(registry string, oauthParams OAuthParams) (accessToken string, err error)
}
```
and for streaming based services like diffcopy, interfaces like:
```
type DiffCopyService interface{
 Proceed(outputFilePackets chan<- FilePacket, inputFileContentRequests <-chan ContentRequest) error
}
```